File: BE/BE/src/main/java/DoAn/BE/project/entity/Project.java
package DoAn.BE.project.entity;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonIgnore;

import jakarta.persistence.*;
import lombok.*;

import DoAn.BE.user.entity.User;
import DoAn.BE.hr.entity.PhongBan;

// Entity quản lý dự án (Agile/Scrum projects)
@Entity
@Table(name = "projects")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Project {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "project_id")
    private Long projectId;

    @Column(nullable = false, length = 255)
    private String name;

    @Column(name = "key_project", nullable = false, unique = true, length = 10)
    private String keyProject;  // VD: PROJ-001, HRM-001

    @Column(columnDefinition = "NVARCHAR(MAX)")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(length = 50, nullable = false)
    private ProjectStatus status = ProjectStatus.ACTIVE;

    @Column(name = "start_date")
    private LocalDate startDate;

    @Column(name = "end_date")
    private LocalDate endDate;

    @ManyToOne
    @JoinColumn(name = "created_by", nullable = false)
    private User createdBy;

    @ManyToOne
    @JoinColumn(name = "phongban_id")
    private PhongBan phongBan;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive = true;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    // Relationships
    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnore
    private List<ProjectMember> members;

    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnore
    private List<Issue> issues;

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    // Helper methods
    public boolean isActive() {
        return this.isActive && this.status == ProjectStatus.ACTIVE;
    }

    public boolean isCompleted() {
        return this.status == ProjectStatus.COMPLETED;
    }

    public boolean isOverdue() {
        return this.status == ProjectStatus.ACTIVE && 
               this.endDate != null && 
               this.endDate.isBefore(LocalDate.now());
    }

    // Enum
    public enum ProjectStatus {
        ACTIVE,      // Đang hoạt động
        ON_HOLD,     // Tạm dừng
        OVERDUE,     // Quá hạn
        COMPLETED,   // Hoàn thành
        CANCELLED    // Đã hủy
    }
}

File: mobile/lib/data/models/project.dart
class Project {
  final int projectId;
  final String name;
  final String keyProject;
  final String description;
  final String status;         // ACTIVE, ON_HOLD, OVERDUE, COMPLETED, CANCELLED
  final String? startDate;
  final String? endDate;
  final double? budget;
  final String? createdByName;
  final String? phongBanName;
  final bool isActive;
  final String? createdAt;
  final String? updatedAt;
  final int? memberCount;
  final int? issueCount;

  Project({
    required this.projectId,
    required this.name,
    required this.keyProject,
    required this.description,
    required this.status,
    this.startDate,
    this.endDate,
    this.budget,
    this.createdByName,
    this.phongBanName,
    this.isActive = true,
    this.createdAt,
    this.updatedAt,
    this.memberCount,
    this.issueCount,
  });

  factory Project.fromJson(Map<String, dynamic> json) {
    return Project(
      projectId: json['projectId'] ?? 0,
      name: json['name'] ?? '',
      keyProject: json['keyProject'] ?? '',
      description: json['description'] ?? '',
      status: json['status'] ?? 'ACTIVE',
      startDate: json['startDate'],
      endDate: json['endDate'],
      budget: json['budget']?.toDouble(),
      createdByName: json['createdBy']?['fullName'] ?? json['createdBy']?['username'],
      phongBanName: json['phongBan']?['tenPhongBan'],
      isActive: json['isActive'] ?? true,
      createdAt: json['createdAt'],
      updatedAt: json['updatedAt'],
      memberCount: json['members']?.length ?? json['memberCount'],
      issueCount: json['issues']?.length ?? json['issueCount'],
    );
  }

  // Helper getters
  String get statusDisplay {
    switch (status) {
      case 'ACTIVE': return 'Đang hoạt động';
      case 'ON_HOLD': return 'Tạm dừng';
      case 'OVERDUE': return 'Quá hạn';
      case 'COMPLETED': return 'Hoàn thành';
      case 'CANCELLED': return 'Đã hủy';
      default: return status;
    }
  }

  bool get isCompleted => status == 'COMPLETED';
  bool get isOverdue => status == 'OVERDUE';
}
